floats/floats.go

```diff
@@ -923,3 +923,23 @@ func Within(s []float64, v float64) int {
 	}
 	return -1
 }
+
+// SumCompensated returns the sum of the elements of the slice calculated with greater
+// accuracy than Sum at the expense of additional computation.
+func SumCompensated(s []float64) float64 {
+	// SumCompensated uses an improved version of Kahan's compensated
+	// summation algorithm proposed by Neumaier.
+	// See https://en.wikipedia.org/wiki/Kahan_summation_algorithm for details.
+	sum := 0.
+	c := 0.
+	for _, x := range s {
+		t := sum + x
```

Comment:
I think this may want to have a caveat comment for future us for if the compiler gets smart enough to optimise these away as @vladimir-ch wrote [here](https://github.com/gonum/gonum/pull/1334#discussion_r431744475). It doesn't now, and as you said the tests will fail if it does, but if we have a comment to watch out, it might be easier to find if it does eventuate.

The spec has [this to say](https://golang.org/ref/spec#Arithmetic_operators):
>An implementation may combine multiple floating-point operations into a single fused operation, possibly across statements, and produce a result that differs from the value obtained by executing and rounding the instructions individually. An explicit floating-point type conversion rounds to the precision of the target type, preventing fusion that would discard that rounding.

Maybe even just adding `t := float64(sum + x)` with that comment to head it off paranoically?

Reply:
Done with a small tweak ;-)